File: Maths/distances.js
/**
 * Différentes fonctions de calcul de distances utiles.
 * @author Noam Zeitoun
 * @version 1.0
 */

/**
 * @description Fonction de calcul d'une distance Euclidienne entre 2 points.
 * @param {number} x1 - Abscisse du premier point.
 * @param {number} y1 - Ordonnée du premier point.
 * @param {number} x2 - Abscisse du deuxième point.
 * @param {number} y2 - Ordonnée du deuxième point.
 * @returns {number} Distance Euclidienne entre les 2 points.
 */
const distanceEuclidienne = (x1, y1, x2, y2) => {
  const width = x2 - x1;
  const height = y2 - y1;
  return Math.sqrt(width * width + height * height);
};

/**
 * @description Fonction de calcul d'une distance de Manhattan entre 2 points.
 * @param {number} x1 - Abscisse du premier point.
 * @param {number} y1 - Ordonnée du premier point.
 * @param {number} x2 - Abscisse du deuxième point.
 * @param {number} y2 - Ordonnée du deuxième point.
 * @returns {number} Distance de Manhattan entre les 2 points.
 */
const distanceManhattan = (x1, y1, x2, y2) => {
  const width = x2 - x1;
  const height = y2 - y1;
  return Math.abs(width) + Math.abs(height);
};

/**
 * @description Fonction de calcul d'une distance de Minkowski entre 2 points.
 * @param {number} x1 - Abscisse du premier point.
 * @param {number} y1 - Ordonnée du premier point.
 * @param {number} x2 - Abscisse du deuxième point.
 * @param {number} y2 - Ordonnée du deuxième point.
 * @param {number} p - Paramètre de la fonction de distance.
 * @returns {number} Distance de Minkowski entre les 2 points.
 */
const distanceMinkowski = (x1, y1, x2, y2, p) => {
  const width = x2 - x1;
  const height = y2 - y1;
  return (Math.abs(width) ** p + Math.abs(height) ** p) ** (1 / p);
};

/**
 * @description Fonction de calcul d'une distance de Tchebychev entre 2 points.
 * @param {number} x1 - Abscisse du premier point.
 * @param {number} y1 - Ordonnée du premier point.
 * @param {number} x2 - Abscisse du deuxième point.
 * @param {number} y2 - Ordonnée du deuxième point.
 * @returns {number} Distance de Tchebychev entre les 2 points.
 */
const distanceTchebychev = (x1, y1, x2, y2) => {
  const width = x2 - x1;
  const height = y2 - y1;
  return Math.max(width, height);
};

/**
 * @description Fonction de calcul d'une distance de Hammings entre 2 strings.
 * @param {string} str1 - Première string.
 * @param {string} str2 - Deuxième string.
 * @returns {number} Distance de Hammings entre les 2 strings.
 */
const distanceHamming = (str1, str2) => {
  let distance = 0;
  for (let i = 0; i < str1.length; i++) {
    if (str1[i] !== str2[i]) {
      distance++;
    }
  }
  return distance;
};

/**
 * @description Fonction de calcul d'une distance de Jaccard entre 2 arrays.
 * @param {array} arr1 - Premier array.
 * @param {array} arr2 - Deuxième array.
 * @returns {number} Distance de Jaccard entre les 2 arrays.
 */
const distanceJaccard = (arr1, arr2) => {
  const union = new Set([...arr1, ...arr2]);
  const intersection = new Set([...arr1].filter((x) => arr2.includes(x)));
  return 1 - intersection.size / union.size;
};

/**
 * @description Fonction de calcul du coefficient de Sørensen-Dice entre 2 arrays.
 * @param {array} arr1 - Premier array.
 * @param {array} arr2 - Deuxième array.
 * @returns {number} Coefficient de Sørensen-Dice entre les 2 arrays.
 */
const distanceSorensenDice = (arr1, arr2) => {
  const union = new Set([...arr1, ...arr2]);
  const intersection = new Set([...arr1].filter((x) => arr2.includes(x)));
  return (2 * intersection.size) / union.size;
};

/**
 * @description Fonction de calcul d'une distance de Levenshtein entre 2 strings.
 * @param {string} str1 - Première string.
 * @param {string} str2 - Deuxième string.
 * @returns {number} Distance de Levenshtein entre les 2 strings.
 */
const distanceLevenshtein = (str1, str2) => {
  const m = str1.length;
  const n = str2.length;
  const matrix = [];
  let i;
  let j;

  if (m === 0) return n;
  if (n === 0) return m;

  for (i = 0; i <= m; i++) {
    matrix[i] = [];
    matrix[i][0] = i;
  }
  for (j = 0; j <= n; j++) {
    matrix[0][j] = j;
  }

  for (i = 1; i <= m; i++) {
    for (j = 1; j <= n; j++) {
      if (str1.charAt(i - 1) === str2.charAt(j - 1)) {
        matrix[i][j] = matrix[i - 1][j - 1];
      } else {
        matrix[i][j] = Math.min(
          matrix[i - 1][j - 1] + 1, // substitution
          Math.min(
            matrix[i][j - 1] + 1, // insertion
            matrix[i - 1][j] + 1 // suppresion
          )
        );
      }
    }
  }
  return matrix[m][n];
};

/**
 * @description Fonction de calcul de la formule de haversine
 * pour déterminer la distance du grand cercle
 * entre deux points d'une sphère,
 * à partir de leurs longitudes et latitudes
 * @param {number} lat1 - Latitude du premier point.
 * @param {number} lon1 - Longitude du premier point.
 * @param {number} lat2 - Latitude du deuxième point.
 * @param {number} lon2 - Longitude du deuxième point.
 * @returns {number} Distance du grand cercle entre les 2 points.
 */
const formuleHaversine = (lat1, lon1, lat2, lon2) => {
  const R = 6371e3; // metres
  const phi1 = (lat1 * Math.PI) / 180; // phi, lambda en radians
  const phi2 = (lat2 * Math.PI) / 180;
  const deltaPhi = ((lat2 - lat1) * Math.PI) / 180;
  const deltaLambda = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(deltaPhi / 2) * Math.sin(deltaPhi / 2) +
    Math.cos(phi1) *
      Math.cos(phi2) *
      Math.sin(deltaLambda / 2) *
      Math.sin(deltaLambda / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

  const result = ((R * c) / 1000).toFixed(2); // en kilomètres
  return result;
};

// Tests
console.table({
  "distanceEuclidienne(0, 0, 1, 1)": distanceEuclidienne(0, 0, 1, 1),
  "distanceManhattan(0, 0, 1, 1)": distanceManhattan(0, 0, 1, 1),
  "distanceMinkowski(0, 0, 1, 1, 1)": distanceMinkowski(0, 0, 1, 1, 1),
  "distanceTchebychev(0, 0, 1, 1)": distanceTchebychev(0, 0, 1, 1),
  "distanceHamming('000', '101')": distanceHamming("000", "101"),
  "distanceJaccard([0, 0, 0], [1, 0, 1])": distanceJaccard(
    [0, 0, 0],
    [1, 0, 1]
  ),
  "distanceSorensenDice([0, 0, 0], [1, 0, 1])": distanceSorensenDice(
    [0, 0, 0],
    [1, 0, 1]
  ),
  "distanceLevenshtein('000', '101')": distanceLevenshtein("000", "101"),
  "formuleHaversine(0, 0, 1, 1)": formuleHaversine(0, 0, 1, 1),
});

File: Python/Maths/distances.py
from math import atan2, cos, radians, sin, sqrt
"""
Description:
    Différentes fonctions de calcul de distances utiles.

Auteur:
    Noam Zeitoun

Version:
    1.0
"""

def distance_euclidienne(x1: float, y1: float, x2: float, y2: float) -> float:
    """
    Description:
        Calcule la distance euclidienne entre deux points.
    
    Paramètres:
        x1: {float} -- Abscisse du premier point.
        y1: {float} -- Ordonnée du premier point.
        x2: {float} -- Abscisse du second point.
        y2: {float} -- Ordonnée du second point.
    
    Retourne:
        {float} -- Distance euclidienne entre les deux points.
    
    Exemple:
        >>> distance_euclidienne(0, 0, 1, 1)
        1.4142135623730951
    """
    width = x2 - x1
    height = y2 - y1
    return sqrt(width**2 + height**2)

def distance_manhattan(x1: float, y1: float, x2: float, y2: float) -> float:
    """
    Description:
        Calcule la distance manhattan entre deux points.
    
    Paramètres:
        x1: {float} -- Abscisse du premier point.
        y1: {float} -- Ordonnée du premier point.
        x2: {float} -- Abscisse du second point.
        y2: {float} -- Ordonnée du second point.
    
    Retourne:
        {float} -- Distance manhattan entre les deux points.
    
    Exemple:
        >>> distance_manhattan(0, 0, 1, 1)
        2.0
    """
    width = x2 - x1
    height = y2 - y1
    return abs(width) + abs(height)

def distance_minkowski(x1: float, y1: float, x2: float, y2: float, p: float) -> float:
    """
    Description:
        Calcule la distance Minkowski entre deux points.
    
    Paramètres:
        x1: {float} -- Abscisse du premier point.
        y1: {float} -- Ordonnée du premier point.
        x2: {float} -- Abscisse du second point.
        y2: {float} -- Ordonnée du second point.
        p: {float} -- Paramètre de la fonction Minkowski.
    
    Retourne:
        {float} -- Distance Minkowski entre les deux points.
    
    Exemple:
        >>> distance_minkowski(0, 0, 1, 1, 1)
        2.0
    """
    width = x2 - x1
    height = y2 - y1
    return (abs(width)**p + abs(height)**p)**(1/p)

def distance_tchebychev(x1: float, y1: float, x2: float, y2: float) -> float:
    """
    Description:
        Calcule la distance Tchebychev entre deux points.
    
    Paramètres:
        x1: {float} -- Abscisse du premier point.
        y1: {float} -- Ordonnée du premier point.
        x2: {float} -- Abscisse du second point.
        y2: {float} -- Ordonnée du second point.
    
    Retourne:
        {float} -- Distance Tchebychev entre les deux points.
    
    Exemple:
        >>> distance_tchebychev(0, 0, 1, 1)
        1.0
    """
    width = x2 - x1
    height = y2 - y1
    return max(abs(width), abs(height))

def distance_hamming(str1: str, str2: str) -> int:
    """
    Description:
        Calcule la distance Hamming entre deux chaînes de caractères.
    
    Paramètres:
        str1: {str} -- Première chaîne de caractères.
        str2: {str} -- Deuxième chaîne de caractères.
    
    Retourne:
        {int} -- Distance Hamming entre les deux chaînes de caractères.
    
    Exemple:
        >>> distance_hamming("000", "101")
        2
    """
    return sum(1 for i in range(len(str1)) if str1[i] != str2[i])

def distance_jaccard(arr1: list, arr2: list) -> float:
    """
    Description:
        Calcule la distance Jaccard entre deux listes.
    
    Paramètres:
        arr1: {list} -- Première liste.
        arr2: {list} -- Deuxième liste.
    
    Retourne:
        {float} -- Distance Jaccard entre les deux listes.
    
    Exemple:
        >>> distance_jaccard([0, 0, 0], [1, 0, 1])
        0.5
    """
    union = set(arr1) | set(arr2)
    intersection = set(arr1) & set(arr2)
    return 1 - len(intersection) / len(union)

def distance_sorensen_dice(arr1: list, arr2: list) -> float:
    """
    Description:
        Calcule la distance Sorensen-Dice entre deux listes.
    
    Paramètres:
        arr1: {list} -- Première liste.
        arr2: {list} -- Deuxième liste.
    
    Retourne:
        {float} -- Distance Sorensen-Dice entre les deux listes.
    
    Exemple:
        >>> distance_sorensen_dice([0, 0, 0], [1, 0, 1])
        1
    """
    union = set(arr1) | set(arr2)
    intersection = set(arr1) & set(arr2)
    return 2 * len(intersection) / len(union)

def distance_levenshtein(str1: str, str2: str) -> int:
    """
    Description:
        Calcule la distance Levenshtein entre deux chaînes de caractères.
    
    Paramètres:
        str1: {str} -- Première chaîne de caractères.
        str2: {str} -- Deuxième chaîne de caractères.
    
    Retourne:
        {int} -- Distance Levenshtein entre les deux chaînes de caractères.
    
    Exemple:
        >>> distance_levenshtein("000", "101")
        2
    """
    if len(str1) < len(str2):
        return distance_levenshtein(str2, str1)
    if len(str2) == 0:
        return len(str1)
    previous_row = range(len(str2) + 1)
    for i, c1 in enumerate(str1):
        current_row = [i + 1]
        for j, c2 in enumerate(str2):
            insertions = previous_row[j + 1] + 1
            deletions = current_row[j] + 1
            substitutions = previous_row[j] + (c1 != c2)
            current_row.append(min(insertions, deletions, substitutions))
        previous_row = current_row
    return previous_row[-1]

def formule_haversine(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Description:
        Calcule la distance entre deux points par la formule de Haversine.
    
    Paramètres:
        lat1: {float} -- Latitude du premier point.
        lon1: {float} -- Longitude du premier point.
        lat2: {float} -- Latitude du second point.
        lon2: {float} -- Longitude du second point.
    
    Retourne:
        {float} -- Distance entre les deux points.
    
    Exemple:
        >>> formule_haversine(0, 0, 1, 1)
        157.24938127194397
    """
    EARTH_RADIUS = 6371e3
    dLat = radians(lat2 - lat1)
    dLon = radians(lon2 - lon1)
    lat1 = radians(lat1)
    lat2 = radians(lat2)
    a = sin(dLat/2)**2 + cos(lat1) * cos(lat2) * sin(dLon/2)**2
    c = 2 * atan2(sqrt(a), sqrt(1-a))
    return (EARTH_RADIUS * c) / 1000

# TESTS
print(
    'distance_euclidienne(0, 0, 1, 1)'+' ---------------------> ', distance_euclidienne(0, 0, 1, 1),
    '\ndistance_manhattan(0, 0, 1, 1)'+' -----------------------> ', distance_manhattan(0, 0, 1, 1),
    '\ndistance_tchebychev(0, 0, 1, 1)'+' ----------------------> ', distance_tchebychev(0, 0, 1, 1),
    '\ndistance_hamming("000", "101")'+' -----------------------> ', distance_hamming("000", "101"),
    '\ndistance_jaccard([0, 0, 0], [1, 0, 1])'+' ---------------> ', distance_jaccard([0, 0, 0], [1, 0, 1]),
    '\ndistance_sorensen_dice([0, 0, 0], [1, 0, 1])'+' ---------> ', distance_sorensen_dice([0, 0, 0], [1, 0, 1]), 
    '\ndistance_levenshtein("000", "101")'+' -------------------> ', distance_levenshtein("000", "101"), 
    '\nformule_haversine(0, 0, 1, 1)'+' ------------------------> ', formule_haversine(0, 0, 1, 1)
)